### PR TITLE
fix bug multi environment on aws sqs/sns #2364

### PR DESCRIPTION
**Pull request by Sbolder**
Signed-off-by: Davide Boldrin <davi1994@hotmail.it>

# Description

I have changed the metod to create a policy on pubsub aws subcription, and now support multiEnvironment.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #2364
https://github.com/dapr/components-contrib/issues/2364
## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X ] Code compiles correctly
* [ X] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pubsub/aws/snssqs/policy.go [hunk 1]
```diff
@@ -74,18 +44,17 @@ type policy struct {
...
-					return true
-				}
+			if s.Condition.ValueArnEquals.AwsSourceArn == snsArn {
+				return true
 			}
+
 			// insert it if it does not exist
-			p.Statement[i].Condition.ForAllValuesArnEquals.AwsSourceArn = append(p.Statement[i].Condition.ForAllValuesArnEquals.AwsSourceArn, snsArn)
-			return false
+			//p.Statement[i].Condition.ForAllValuesArnEquals.AwsSourceArn = append(p.Statement[i].Condition.ForAllValuesArnEquals.AwsSourceArn, snsArn)
```
Comment (ItalyPaleAle) [inline pubsub/aws/snssqs/policy.go:56]: Please remove code that's now unused, rather than just commenting it out.

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]
Ciao Davide, thanks for this PR.

Can you please explain how this fixes the issue you've reported?

Also, please fix the DCO issue: https://docs.dapr.io/contributing/contributing-overview/#developer-certificate-of-origin-signing-your-work

**Comment by Sbolder** [comment]
@ItalyPaleAle 
 in the previous version the method "tryInsertCondition" try to create one statement for each resource, and append a new condition for each new sns, if you have one aws account with two environment this generate a bad request on startup of second service.

With my fix, now the method create one statement for each resource so, the relation until sqs ans sns is 1..1 on policy, and is more clear to understand.
Otherwhise if you have for example:
on env "dev" one publisher and one subscriber
on env "test" one publisher and one subscriber
with my fix each services write their policy and the second service dosn't change other service policy, is more linear and clear.

please feel free to contact me for any questions.
D.

**Comment by berndverst** [comment]
Will consider merging this after we merge the SNS/SQS certification test. I don't know enough about this change so will need input from community members with more knowledge of SNS/SQS

**Comment by yaron2** [comment]
cc @amimo

**Comment by amimimor** [comment]
> #2364

@mukundansundar, @yaron2, others, please correct me if I'm wrong: issue #2364 describes a feature that can be summarized as a "Fan-In" pattern, meaning: two producer topics (2 SNSs) possibly originating from different environments (2 EKSs), and one subscriber queue (SQS, possibly in a different cluster). 

Just to clarify, currently, the AWS SNS/SQS PubSub component supports either an at-least 1:1 ratio of sns-topic:sqs-queue, which *possibly can support Fan-In* (if many publisher apps push messages to a given SNS topic), or at-least 1:many (many on the consuming side), using the component default behavior (it would connect 1 SNS topic to many SQSs, each dedicated per Dapr consuming app). The problem that #2364 wants to solve is therefore quite unique, where many publisher apps,  want to push messages to a queue by using different topics - which I fail to understand why does one need 2 or more topics to send messages that would eventually reach a single queue, if they could send, from many producer apps, messages to a given topic.

From my experience, the way we use EKS and AWS environments in general, we would never allow cross-EKS resource sharing and *personally* I am strongly against encouraging behaviors that might, due to lack of awareness/knowledge, result in the creation of a security vulnerability - e.g. data leakage from one EKS cluster to another. For that reason, if the user wishes to do whatever they want, we have introduced the ability to disable the automatic provisioning and configuration of resources and their security policies.

**Comment by yaron2** [comment]
cc @Sbolder

**Comment by Sbolder** [comment]
Hello  @yaron2 , @amimimor 

I think I didn't explain myself well.
I'll try to summarize my pull request.

The requirement is to have two EKS clusters on the same account (eks-develop and eks-test).
Now the problem arises when I deploy the first environment, the following are created:
1 sqs (writer)
1 sns (reader-dev)
with all the policies created correctly.

When I deploy the second test environment, it is recreated (or maybe not because it already exists):
1 sqs (writer) with the same name as the entity managed by Dapr
1 sns (reader-test)

The problem at this point is that the policies are generated incorrectly, not allowing both sns to subscribe to the sqs, but only one of them is able to do it and it works.

I understand the concern that there may be a exchange of messages between different EKS clusters (something to be avoided at all costs), but as far as I understand how Dapr works, it always reads only from the sns queue that it creates at startup, so I don't think a exchange of messages between different EKS clusters is possible.

I hope I was clearer.

**Pull request merged**